audioipc/src/rpccore.rs

```diff
@@ -45,62 +137,46 @@ pub trait Server {
 }
 
 // RPC Client Proxy implementation.
-type ProxyKey = usize;
-type ProxyRequest<Request> = (ProxyKey, Request);
+type ProxyRequest<Request, Response> = (Request, CompletionWriter<Response>);
 
 // RPC Proxy that may be `clone`d for use by multiple owners/threads.
 // A Proxy `call` arranges for the supplied request to be transmitted
 // to the associated Server via RPC and blocks awaiting the response
-// via `response_rx`.
-// A Proxy is associated with the ClientHandler via `handler_tx` to send requests,
-// `response_rx` to receive responses, and uses `key` to identify the Proxy with
-// the sending side of `response_rx`  ClientHandler.
-// Each Proxy is registered with the ClientHandler on initialization via the
-// ProxyManager and unregistered when dropped.
+// via the associated `Completion`.
 // A ClientHandler normally lives until the last Proxy is dropped, but if the ClientHandler
-// encounters an internal error, `handler_tx` and `response_tx` will be closed.
+// encounters an internal error, `requests` will fail to upgrade, allowing
```

Comment:
Does it mean the Proxy could outlive ClientHandler sometimes? (In that case, does EventLoop remove a `Connection` while one of the `Connection`'s corresponding proxies is alive?) When will this happen? Is it possible to write a test for it?

Reply:
Yep, a Connection will be removed it if experiences an internal error (e.g. remote spuriously disconnected), causing the ClientHandler to be destroyed.  Any Proxy may outlive this happening, but in-flight or subsequent Proxy operations should fail immediately - this is detected and handled by the upgrade of the `Weak` reference to the associated `ArrayQueue` failing.

Reply:
This should already be covered by `test_dead_server` in ipccore.rs, but please let me know if you think an additional test is still necessary.